src/compiler/scala/tools/nsc/transform/UnCurry.scala

```diff
@@ -744,72 +744,88 @@ abstract class UnCurry extends InfoTransform
+            (oldTps, newTps)
+
+          case _ => (Nil, Nil)
         }
-        val seqargs = map2(locals, forwParams) {
-          case (null, argsym) => Ident(argsym)
-          case (l, _)         => l
+
+        def toArrayType(tp: Type, newParam: Symbol): Type = {
+          val arg = elementType(SeqClass, tp)
+          val elem = if (arg.typeSymbol.isTypeParameterOrSkolem && !(arg <:< AnyRefTpe)) {
+            // To prevent generation of an `Object` parameter from `Array[T]` parameter later
+            // as this would crash the Java compiler which expects an `Object[]` array for varargs
+            //   e.g.        def foo[T](a: Int, b: T*)
+            //   becomes     def foo[T](a: Int, b: Array[Object])
+            //   instead of  def foo[T](a: Int, b: Array[T]) ===> def foo[T](a: Int, b: Object)
+            //
+            // In order for the forwarder method to type check we need to insert a cast
+            //   def foo'[T'](a: Int, b: Array[Object]) = foo[T'](a, wrapRefArray(b).asInstanceOf[Seq[T']])
+            // The target type for that cast is stored in the `arrayTypesMappedToObject` map.
+            val originalArg = arg.substSym(oldTps, tps)
+            arrayTypesMappedToObject(newParam) = originalArg
+            // Store the type parameter that was replaced by Object to emit the correct generic signature
+            newParam.updateAttachment(new erasure.TypeParamVarargsAttachment(originalArg))
+            ObjectTpe
+          } else
+            arg
+          arrayType(elem)
         }
-        val end = if (forwsym.isConstructor) List(UNIT) else Nil
 
-        DefDef(forwsym, BLOCK(Apply(gen.mkAttributedRef(flatdd.symbol), seqargs) :: end : _*))
+        val ps = map2(oldPs, isRepeated)((oldParam, isRep) => {
+          val newParam = oldParam.cloneSymbol(forwSym)
+          val tp = if (isRep) toArrayType(oldParam.tpe, newParam) else oldParam.tpe
+          newParam.setInfo(tp)
+        })
+
+        val resTp = dd.symbol.tpe_*.finalResultType.substSym(oldPs, ps)
+        val mt = MethodType(ps, resTp)
+        val r = if (tps.isEmpty) mt else PolyType(tps, mt)
+        r.substSym(oldTps, tps)
+      }
+
+      forwSym.setInfo(forwTpe)
+      val newPs = forwTpe.params
+
+      val theTyper = typer.atOwner(dd, currentClass)
+      val forwTree = theTyper.typedPos(dd.pos) {
+        val seqArgs = map3(newPs, oldPs, isRepeated)((param, oldParam, isRep) => {
+          if (!isRep) Ident(param)
+          else {
+            val parTp = elementType(ArrayClass, param.tpe)
+            val wrap = gen.mkWrapArray(Ident(param), parTp)
+            arrayTypesMappedToObject.get(param) match {
+              case Some(tp) => gen.mkCast(wrap, seqType(tp))
+              case _ => wrap
+            }
+          }
+        })
+        val end = if (forwSym.isConstructor) List(UNIT) else Nil
+
+        DefDef(forwSym, BLOCK(Apply(gen.mkAttributedRef(flatdd.symbol), seqArgs) :: end : _*))
```

Comment:
This pattern of block with singleton stat + unit expr is pretty common. At some point we should introduce a factory method for it (I'm not saying in this PR -- some day)